### PR TITLE
More wildcards in CORS when used without credentials

### LINKED ISSUE #251: Allow * for Access-Control-Allow-Headers and Access-Control-Allow-Methods
This has almost certainly been discussed before, but would it be possible to allow * (allow-all) as a separate value for the **Access-Control-Allow-Headers** CORS response header?

This would allow _all_ non-simple headers passed in the request to be added to the browser's preflight cache. This si currently possible by simply mirroring back the value of the **Access-Control-Request-Headers** request header, but this would be much simpler. The browser would need to track the request headers passed and add them all to their preflight cache (rather than simply parse them out from the **Access-Control-Allow-Headers** response header, assuming that's what they currently do), but that's not too hard to do.

So the spec would become the following:

Access-Control-Allow-Headers = "Access-Control-Allow-Headers" ":" #field-name | "*"

### LINKED ISSUE #252: Allow * for Access-Control-Expose-Headers
This has almost certainly been discussed before, but would it be possible to allow * (allow-all) as a separate value for the **Access-Control-Expose-Headers** CORS response header?

This would allow all non-simple headers passed in the request to be 'exposable' to client-side code.

When the original CORS spec was written, there was an assumption (I assume!) that the preflight OPTIONS response would be served by the same application that would serve the subsequent CORS GET/POST/whatever request. Therefore, that application would have full knowledge of what headers the API might eventually respond with, and which of them should be exposed or suppressed.

However, with the advent of CDN's and load balancers, it may be that the code that responds to the preflight OPTIONS request is **not** the actual application that will process the subsequent GET/POST/whatever request. Indeed, website owners might want the OPTIONS response to be served from non-application code, because that code may be quicker to respond to the client.

(Full disclosure: I work for Akamai, a large CDN. Whilst my views don't represent any official Akamai position, we would obviously like this, since it would make our (and out customer's) lives easier. I have also implemented CORS solutions using an F5 LTM load balancer, and run into this problem).

So the spec would become the following:

**Access-Control-Expose-Headers = "Access-Control-Expose-Headers" ":" #field-name | "*"**

If Access-Control-Expose-Headers: * is returned, browser would allow all headers to be returned using e.g. XMLHttpRequest getResponseHeader().

### PR DESCRIPTION
**Pull request by annevk**
Fixes #251 and fixes #252.

### CODE REVIEW

**Comment by annevk** [comment]
@tyoshino review appreciated. @roryhewitt, @craigfrancis, @mozfreddyb, you might be interested in reviewing too. @sicking, I included your suggestion around `Authorization`.

File: Overview.html [hunk 1]
```diff
@@ -1856,12 +1858,14 @@ <h4 id="http-new-header-syntax"><span class="secno">4.2.4 </span>HTTP new-header
 <pre>Access-Control-Request-Method    = <a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.1.1">method</a>
 Access-Control-Request-Headers   = #<a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.2">field-name</a>
 
-Access-Control-Allow-Origin      = origin-or-null / "*"
+wilrdcard                        = "*"
```
Comment (MichaelKohler) [inline Overview.html]: I highly suspect this to be a typo and it should read "wildcard"?

File: Overview.src.html [hunk 2]
```diff
@@ -1785,12 +1787,14 @@ <h4 id=http-new-header-syntax>HTTP new-header syntax</h4>
 <pre>Access-Control-Request-Method    = <span data-anolis-spec=http>method</span>
 Access-Control-Request-Headers   = #<span data-anolis-spec=http>field-name</span>
 
-Access-Control-Allow-Origin      = origin-or-null / "*"
+wilrdcard                        = "*"
```
Comment (MichaelKohler) [inline Overview.src.html]: typo in "wilrdcard"

Comment (annevk) [inline Overview.src.html]: Thank you, fixed.

**Comment by annevk** [comment]
I thought a bit about this and I think one thing we should probably change is to make the CORS-preflight cache support wildcards directly. That could have a pretty dramatic impact on reducing chattiness.

**Comment by roryhewitt** [comment]
@annevk that would be nice. +1 from me.

File: Overview.html [hunk 3]
```diff
@@ -1857,12 +1859,14 @@ <h4 id="http-new-header-syntax"><span class="secno">4.2.4 </span>HTTP new-header
...
+Access-Control-Allow-Origin      = origin-or-null / wildcard
 Access-Control-Allow-Credentials = %x74.72.75.65 ; "true", case-sensitive
-Access-Control-Expose-Headers    = #<a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.2">field-name</a>
+Access-Control-Expose-Headers    = field-name-or-wildcard
```
Comment (roryhewitt) [inline Overview.html]: Shouldn't this be **#field-name-or-wildcard** (with a leading #?)

File: Overview.html [hunk 4]
```diff
@@ -1857,12 +1859,14 @@ <h4 id="http-new-header-syntax"><span class="secno">4.2.4 </span>HTTP new-header
...
+field-name-or-wildcard           = <a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.2">field-name</a> / wildcard
 Access-Control-Max-Age           = <a class="external" data-anolis-spec="http-caching" href="https://tools.ietf.org/html/rfc7234#section-1.2.1">delta-seconds</a>
-Access-Control-Allow-Methods     = #<a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.1.1">method</a>
-Access-Control-Allow-Headers     = #<a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.2">field-name</a></pre>
+Access-Control-Allow-Methods     = #<a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.1.1">method</a> / wildcard
+Access-Control-Allow-Headers     = #field-name-or-wildcard</pre>
```
Comment (roryhewitt) [inline Overview.html]: Should probably add back link to IETF?

File: Overview.html [hunk 5]
```diff
@@ -3644,6 +3683,18 @@ <h3 id="cors-preflight-fetch"><span class="secno">5.7 </span>CORS-preflight fetc
    <a href="#concept-network-error" title="concept-network-error">network error</a>.
 
    <li><p>If one of <var>request</var>'s
+   <a href="#concept-request-header-list" title="concept-request-header-list">header list</a>'s
+   <a href="#concept-header-name" title="concept-header-name">names</a> is a
+   <a href="#cors-non-wildcard-request-header-name">CORS non-wildcard request-header name</a> and is not in <var>headerNames</var>, then
+   return a <a href="#concept-network-error" title="concept-network-error">network error</a>.
+
+   <li><p>If <var>headerNames</var> contains `<code>*</code>`, then remove all occurances of
```
Comment (vigneshshanmugam) [inline Overview.html]: occurrences?

File: Overview.html [hunk 6]
```diff
@@ -1857,12 +1859,14 @@ <h4 id="http-new-header-syntax"><span class="secno">4.2.4 </span>HTTP new-header
 <pre>Access-Control-Request-Method    = <a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.1.1">method</a>
 Access-Control-Request-Headers   = #<a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.2">field-name</a>
 
-Access-Control-Allow-Origin      = origin-or-null / "*"
+wildcard                         = "*"
+Access-Control-Allow-Origin      = origin-or-null / wildcard
 Access-Control-Allow-Credentials = %x74.72.75.65 ; "true", case-sensitive
-Access-Control-Expose-Headers    = #<a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.2">field-name</a>
+Access-Control-Expose-Headers    = field-name-or-wildcard
+field-name-or-wildcard           = <a class="external" data-anolis-spec="http" href="https://tools.ietf.org/html/rfc7230#section-3.2">field-name</a> / wildcard
```
Comment (tyoshino) [inline Overview.html]: Shouldn't this be:
#field-name / wildcard

File: Overview.html (hunk 4 above)
Comment (tyoshino) [inline Overview.html]: and shouldn't this be just field-name-or-wildcard (no hash)?

File: Overview.html [hunk 7]
```diff
@@ -2452,13 +2463,29 @@ <h3 id="main-fetch"><span class="secno">5.1 </span>Main fetch</h3>
...
   <ol>
-   <li><p>If <var>request</var>'s
-   <a href="#concept-request-response-tainting" title="concept-request-response-tainting">response tainting</a> is
-   "<code title="">cors</code>", then set <var>response</var>'s
-   <a href="#concept-response-cors-exposed-header-name-list" title="concept-response-cors-exposed-header-name-list">CORS-exposed header-name list</a>
-   to the result of <a href="#concept-header-parse" title="concept-header-parse">parsing</a>
-   `<a href="#http-access-control-expose-headers"><code title="http-access-control-expose-headers">Access-Control-Expose-Headers</code></a>` in
-   <var>response</var>'s <a href="#concept-response-header-list" title="concept-response-header-list">header list</a>.
+   <li>
+   	<p>If <var>request</var>'s
```
Comment (tyoshino) [inline Overview.html]: broken indentation

File: Overview.html [hunk 8]
```diff
@@ -2452,13 +2463,29 @@ <h3 id="main-fetch"><span class="secno">5.1 </span>Main fetch</h3>
...
+    <a href="#concept-request-response-tainting" title="concept-request-response-tainting">response tainting</a> is
+    "<code title="">cors</code>", then run these substeps:</p>
+
+    <ol>
+     <li><p>Let <var>headerNames</var> be the result <a href="#concept-header-parse" title="concept-header-parse">parsing</a>
```
Comment (tyoshino) [inline Overview.html]: the result of parsing

File: Overview.html (hunk 4 above)
Comment (tyoshino) [inline Overview.html]: ah, is this intended to allow putting * and Authorization together?

Comment (annevk) [inline Overview.html]: Yeah.

File: Overview.html (hunk 6 above)
Comment (annevk) [inline Overview.html]: Yeah, it probably doesn't make sense to allow either here.

File: Overview.html (hunk 3 above)
Comment (annevk) [inline Overview.html]: It should have the leading #, but since this is not an IETF production this should not link.

File: Overview.html (hunk 5 above)
Comment (annevk) [inline Overview.html]: Thank you!

**Comment by annevk** [comment]
I now changed it so that the CORS-preflight cache can store wildcards which should reduce chattiness a whole lot for non-credentialed requests (at the point user agents implement this, of course).

**Comment by annevk** [comment]
@roryhewitt @tyoshino would you be so kind to look over the latest changes? I think this is ready, but I just want to be sure.

**Comment by roryhewitt** [comment]
@annevk I think the second example in section 4.2.6 should have the last line changed to something like this:

> This is because bar.invalid needs to explicitly share each header by listing their names in the `Access-Control-Expose-Headers` response header. Alternatively, bar.invalid could have returned `Access-Control-Expose-Headers: *` which would share all response headers.

Basically showing developers that they can specify `Access-Control-Expose-Headers: *`.

**Comment by annevk** [comment]
Ended up writing an extra paragraph.

**Comment by roryhewitt** [comment]
@annevk Perfect!

File: Overview.src.html [hunk 9]
```diff
@@ -2380,13 +2401,30 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
...
+   <li>
+    <p>If <var>request</var>'s
+    <span title=concept-request-response-tainting>response tainting</span> is
+    "<code title>cors</code>", then run these substeps:</p>
+
+    <ol>
+     <li><p>Let <var>headerNames</var> be the result of
+     <span title=concept-header-parse>parsing</span>
+     `<code title=http-access-control-expose-headers>Access-Control-Expose-Headers</code>` in
+     <var>response</var>'s <span title=concept-response-header-list>header list</span>.
```
Comment (tyoshino) [inline Overview.src.html]: paragraph not enclosed

**Comment by tyoshino** [comment]
Sorry for delay.

lgtm but I want to make sure that it's fine that
- a user-agent may end up sending no credential even when credentials mode is set to include
- in such case, the server cannot know that the client is using `credentials mode == include` only by inspecting the received request, and therefore it's not safe to use the wildcard.

This means that developers are basically required to coordinate use of the wildcard between client code and server. Right?

**Comment by annevk** [comment]
@tyoshino yeah, I think that coordination has effectively always been required (it also applies to the wildcard case for origin). It might help if we start communicating that more clearly, but that would be a new request header and a separate discussion.

**Comment by tyoshino** [comment]
ok. lgtm

**Comment by annevk** [comment]
Thank you (and the others)!

**Comment by roryhewitt** [comment]
@annevk forgive my ignorance, but I'm not sure how this all goes together...

If these new wildcard values are now in the master branch (and are what I see when I go to https://fetch.spec.whatwg.org/#http-new-header-syntax), does that mean that they are **officially supported** by all (major) browsers? Or is this just a new version of the Fetch spec, and as browsers add support for them, they will publicize this fact?

**Comment by annevk** [comment]
@roryhewitt the latter. I wish it were that easy!

**Comment by tonyjin** [comment]
Also not sure how this all goes together, but I noticed the following two don't match:
https://www.w3.org/TR/cors/#access-control-allow-headers-response-header
https://fetch.spec.whatwg.org/#http-new-header-syntax

Specifically:
Access-Control-Allow-Headers: "Access-Control-Allow-Headers" ":" #field-name (in W3)
Access-Control-Allow-Headers     = #field-name-or-wildcard (in Fetch Spec)

As I was writing this I realized the W3 recommendation was published in Jan 2014, so I guess we'll have to wait for a newer version to be published?

Does the latest fetch spec's definition of Access-Control-Allow-Headers supersede the W3 recommendation?

**Comment by annevk** [comment]
@tonyjin yes it does. W3C's copy of CORS is obsolete and should not be used. Unfortunately they have so far refused to clearly mark it as such.

**Comment by roryhewitt** [comment]
Hey Tony,

Yes, the fetch spec on WHATWG (including the CORS headers) supersedes the
W3C spec.

Unfortunately, it appears to be a low priority to get this 'publicized' by,
for instance, updating the W3C spec to make it clear.

Note also that, although the latest fetch spec includes wildcards for the
ACAM, ACAH and ACEH headers, I'm not sure whether these special values are
actually supported by any browsers yet...

Perhaps @annevk can comment?

Rory

**Comment by tonyjin** [comment]
Thanks for clarifying and working on this @annevk & @roryhewitt - I ran into this PR after getting frustrated at adding Access-Control-Allow-Headers: [a bajillion headers] in nginx 🙄

**Pull request merged**